File: built-in plugins dsp/mb-mix-console.cpp
/**
 * MB Console Emulator
 * Category : effect
 * Type     : mixing
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Analog console channel strip emulation with subtle harmonic saturation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIX_CONSOLE_H
#define MB_MIX_CONSOLE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMixConsole : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mix-console";
    static constexpr const char* PLUGIN_NAME    = "MB Console Emulator";
    static constexpr const char* PLUGIN_TYPE    = "mixing";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.3f;  // range [0, 1]
    float crosstalk = 0.2f;  // range [0, 1]
    float hiss = 0.1f;  // range [0, 1]
    float output = 0.8f;  // range [0, 1]
    };

    MbMixConsole() = default;
    ~MbMixConsole() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.crosstalk = std::clamp(params.crosstalk, 0f, 1f);
        params.hiss = std::clamp(params.hiss, 0f, 1f);
        params.output = std::clamp(params.output, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Console Emulator
        return input;
    }
};

#endif // MB_MIX_CONSOLE_H

File: built-in plugins dsp/mb-mix-console.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMixConsolePlugin: PluginDefinition = { id: 'mb-mix-console', slug: 'mb-mix-console', name: 'MB Console Emulator', category: 'effect', type: 'mixing' as any, version: '1.0.0', description: 'Analog console channel strip emulation with subtle harmonic saturation', author: 'Max Booster', grade: 'A', parameters: [{ id: 'drive', name: 'Drive', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'crosstalk', name: 'Crosstalk', type: 'float', defaultValue: 0.2, minValue: 0, maxValue: 1, automatable: true }, { id: 'hiss', name: 'Hiss', type: 'float', defaultValue: 0.1, minValue: 0, maxValue: 1, automatable: true }, { id: 'output', name: 'Output', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { drive: 0.3, crosstalk: 0.2, hiss: 0.1, output: 0.8 } };

export default MbMixConsolePlugin;
